desidlas/training/model.py

```diff
@@ -0,0 +1,304 @@
+    inputsize_fc1 = int(math.ceil(math.ceil(math.ceil(math.ceil(math.ceil(math.ceil(
+        INPUT_SIZE / conv1_stride) / pool1_stride) / conv2_stride) / pool2_stride) / conv3_stride) / pool3_stride)) * conv3_filters
+    # batch_size = x.get_shape().as_list()[0]
+    h_pool3_flat = tf.reshape(h_pool3, [-1, inputsize_fc1])
+
+    W_fc1 = weight_variable([inputsize_fc1, fc1_n_neurons])
+    b_fc1 = bias_variable([fc1_n_neurons])
+    h_fc1 = tf.nn.relu(tf.matmul(h_pool3_flat, W_fc1) + b_fc1)
+
+    # Dropout FC1
+    h_fc1_drop = tf.nn.dropout(h_fc1, 1 - (keep_prob))
+
+    W_fc2_1 = weight_variable([fc1_n_neurons, fc2_1_n_neurons])
+    b_fc2_1 = bias_variable([fc2_1_n_neurons])
+    W_fc2_2 = weight_variable([fc1_n_neurons, fc2_2_n_neurons])
+    b_fc2_2 = bias_variable([fc2_2_n_neurons])
+    W_fc2_3 = weight_variable([fc1_n_neurons, fc2_3_n_neurons])
+    b_fc2_3 = bias_variable([fc2_3_n_neurons])
+
+    # FC2 activations
+    h_fc2_1 = tf.nn.relu(tf.matmul(h_fc1_drop, W_fc2_1) + b_fc2_1)
+    h_fc2_2 = tf.nn.relu(tf.matmul(h_fc1_drop, W_fc2_2) + b_fc2_2)
+    h_fc2_3 = tf.nn.relu(tf.matmul(h_fc1_drop, W_fc2_3) + b_fc2_3)
+
+    # FC2 Dropout [1-3]
+    h_fc2_1_drop = tf.nn.dropout(h_fc2_1, 1 - (keep_prob))
+    h_fc2_2_drop = tf.nn.dropout(h_fc2_2, 1 - (keep_prob))
+    h_fc2_3_drop = tf.nn.dropout(h_fc2_3, 1 - (keep_prob))
+
+    # Readout Layer
+    W_fc3_1 = weight_variable([fc2_1_n_neurons, 1])
+    b_fc3_1 = bias_variable([1])
+    W_fc3_2 = weight_variable([fc2_2_n_neurons, 1])
+    b_fc3_2 = bias_variable([1])
+    W_fc3_3 = weight_variable([fc2_3_n_neurons, 1])
+    b_fc3_3 = bias_variable([1])
+
+    # y_fc4 = tf.add(tf.matmul(h_fc3_drop, W_fc4), b_fc4)
+    # y_nn = tf.reshape(y_fc4, [-1])
+    y_fc4_1 = tf.add(tf.matmul(h_fc2_1_drop, W_fc3_1), b_fc3_1)
+    y_nn_classifier = tf.reshape(y_fc4_1, [-1], name='y_nn_classifer')
+    y_fc4_2 = tf.add(tf.matmul(h_fc2_2_drop, W_fc3_2), b_fc3_2)
+    y_nn_offset = tf.reshape(y_fc4_2, [-1], name='y_nn_offset')
+    y_fc4_3 = tf.add(tf.matmul(h_fc2_3_drop, W_fc3_3), b_fc3_3)
+    y_nn_coldensity = tf.reshape(y_fc4_3, [-1], name='y_nn_coldensity')
+
+    # Train and Evaluate the model
+    loss_classifier = tf.add(tf.nn.sigmoid_cross_entropy_with_logits(logits=y_nn_classifier, labels=label_classifier),
+                             l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                          tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_1)),
+                             name='loss_classifier')
+    loss_offset_regression = tf.add(tf.reduce_sum(input_tensor=tf.nn.l2_loss(y_nn_offset - label_offset)),
+                                    l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                                 tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_2)),
+                                    name='loss_offset_regression')
+    epsilon = 1e-6
```

Comment:
This should probably be a parameter in the method call.

Reply:
This value is used in the loss function of column density, Parks chose this value because he said this is a small value safe from 32-bit floating point rounding error.